File: springboot/springboot/src/main/java/com/joa/springboot/PuntoDeVenta/PuntoDeVentaService.java
package com.joa.springboot.PuntoDeVenta;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.joa.springboot.Boliche.Boliche;
import com.joa.springboot.Boliche.BolicheRepository;
import com.joa.springboot.Usuario.Usuario;
import com.joa.springboot.Usuario.UsuarioRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class PuntoDeVentaService {

    @Autowired
    private PuntoDeVentaRepository puntoDeVentaRepository;

    @Autowired
    private BolicheRepository bolicheRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    // ✅ Crear un nuevo Punto de Venta
    public PuntoDeVentaResponseDTO createPuntoDeVenta(PuntoDeVentaRequestDTO requestDTO) {
        Boliche boliche = bolicheRepository.findById(requestDTO.getBolicheId())
                .orElseThrow(() -> new RuntimeException("Boliche no encontrado"));

        Usuario seguridad = usuarioRepository.findById(requestDTO.getSeguridadId())
                .orElseThrow(() -> new RuntimeException("Usuario de seguridad no encontrado"));

        PuntoDeVenta puntoDeVenta = new PuntoDeVenta(requestDTO.getNombre(), boliche, seguridad);
        puntoDeVenta = puntoDeVentaRepository.save(puntoDeVenta);

        return mapToDTO(puntoDeVenta);
    }

    // ✅ Obtener todos los puntos de venta
    public List<PuntoDeVentaResponseDTO> getAllPuntosDeVenta() {
        return puntoDeVentaRepository.findAll().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    // ✅ Obtener un punto de venta por ID
    public PuntoDeVentaResponseDTO getPuntoDeVentaById(Long id) {
        return puntoDeVentaRepository.findById(id)
                .map(this::mapToDTO)
                .orElseThrow(() -> new RuntimeException("Punto de Venta no encontrado"));
    }

    // ✅ Obtener puntos de venta por boliche
    public List<PuntoDeVentaResponseDTO> getPuntosDeVentaByBoliche(Long bolicheId) {
        return puntoDeVentaRepository.findByBolicheId(bolicheId).stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    // ✅ Eliminar un punto de venta
    public void deletePuntoDeVenta(Long id) {
        puntoDeVentaRepository.deleteById(id);
    }

    // Método de conversión a DTO
    private PuntoDeVentaResponseDTO mapToDTO(PuntoDeVenta puntoDeVenta) {
        return new PuntoDeVentaResponseDTO(
                puntoDeVenta.getId(),
                puntoDeVenta.getNombre(),
                puntoDeVenta.getBoliche().getNombre(),
                puntoDeVenta.getSeguridad().getUsername()
        );
    }
}

File: angular/src/app/core/services/punto-de-venta.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { PuntoDeVenta, PuntoDeVentaRequest } from '../model/punto-de-venta.model';

@Injectable({
  providedIn: 'root',
})
export class PuntoDeVentaService {
  private apiUrl = `${environment.apiUrl}/api/puntos-de-venta`;

  constructor(private http: HttpClient) {}

  // Obtener todos los puntos de venta de un boliche
  getPuntosDeVenta(bolicheId: number): Observable<PuntoDeVenta[]> {
    return this.http.get<PuntoDeVenta[]>(`${this.apiUrl}/boliche/${bolicheId}`);
  }

  // Crear un nuevo punto de venta
  createPuntoDeVenta(puntoDeVenta: PuntoDeVentaRequest): Observable<PuntoDeVenta> {
    return this.http.post<PuntoDeVenta>(this.apiUrl, puntoDeVenta);
  }

  // Eliminar un punto de venta
  deletePuntoDeVenta(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Obtener todos los puntos de venta de un boliche
  getPuntosDeVentaByBoliche(bolicheId: number): Observable<PuntoDeVenta[]> {
    return this.http.get<PuntoDeVenta[]>(`${this.apiUrl}/boliche/${bolicheId}`);
  }


}